state-synchronizer/src/coordinator.rs

```diff
@@ -189,84 +204,124 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
-        self.known_version = self
+    /// Sync up coordinator state with the local storage.
+    async fn sync_state_with_local_storage(&mut self) {
+        self.local_state = self
             .executor_proxy
-            .get_latest_version()
+            .get_local_storage_state()
             .await
-            .expect("[state sync] failed to fetch latest version from storage");
+            .expect("[state sync] failed to sync with local storage");
+    }
+
+    /// The highest available version in the local storage (even if it's not covered by the LI).
+    fn highest_version_in_local_storage(&self) -> u64 {
+        std::cmp::max(
+            self.local_state.highest_local_li.ledger_info().version(),
+            self.local_state.highest_committed_version,
+        )
+    }
+
+    /// In case there has been another pending request it's going to be overridden.
+    /// The caller will be notified about request completion via request.callback oneshot:
+    /// at that moment it's guaranteed that the highest LI exposed by the storage is equal to the
+    /// target LI.
+    /// StateSynchronizer assumes that it's the only one modifying the storage (consensus is not
+    /// trying to commit transactions concurrently).
+    async fn request_sync(&mut self, request: SyncRequest) -> Result<()> {
+        ensure!(
+            self.role == RoleType::Validator,
+            "Sync requests are for validator nodes only"
+        );
+
+        self.sync_state_with_local_storage().await;
+        let highest_local_li = self.local_state.highest_local_li.ledger_info();
         let target_version = request.target.ledger_info().version();
         counters::TARGET_VERSION.set(target_version as i64);
         debug!(
-            "[state sync] sync requested. Known version: {}, requested_version: {}",
-            self.known_version, target_version
+            "[state sync] sync requested. Known LI: {}, requested_version: {}",
+            highest_local_li, target_version
         );
 
-        if target_version <= self.known_version {
-            debug!("[state sync] sync contains only empty blocks");
-            self.store_transactions(
-                TransactionListWithProof::new_empty(),
-                request.target.clone(),
-            )
-            .await
-            .expect("[state sync] failed to execute empty blocks");
-            if request.callback.send(Ok(())).is_err() {
-                error!("[state sync] coordinator failed to notify subscriber");
-            }
-            return;
+        if target_version < highest_local_li.version() {
+            error!("[state sync] Sync request for an old version");
+
+            request
+                .callback
+                .send(Err(format_err!("Sync request for an old version")))
+                .map_err(|_| format_err!("Callback error"))?;
+            bail!("Sync request for an old version");
+        }
+        if target_version == highest_local_li.version()
```

Comment:
May be we can get rid of this case at all (if the PR for block metadata transactions is in)